File: DoNExt/app/src/main/java/com/wismna/geoffroy/donext/fragments/DynamicDialogFragment.java
package com.wismna.geoffroy.donext.fragments;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;

import com.wismna.geoffroy.donext.R;

/**
 * Created by wismna on 2017-03-21.
 * Sub-class this class to create a dynamic fragment that will act as a Dialog in large layouts and
 *  a full screen fragment in smaller layouts.
 */

public abstract class DynamicDialogFragment extends DialogFragment {
    private View mDialogView = null;
    protected boolean mHasNeutralButton = false;
    protected boolean mIsLargeLayout = false;
    protected Fragment mContentFragment = new Fragment();

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        // This part is only needed on small layouts (large layouts use onCreateDialog)
        if (!mIsLargeLayout) {
            View view = inflater.inflate(R.layout.fragment_dynamic_dialog, container, false);
            AppCompatActivity activity = (AppCompatActivity) getActivity();
            activity.setSupportActionBar(setToolbarTitle(view));

            ActionBar actionBar = activity.getSupportActionBar();
            if (actionBar != null) {
                actionBar.setDisplayHomeAsUpEnabled(true);
                actionBar.setHomeButtonEnabled(true);
                actionBar.setHomeAsUpIndicator(android.R.drawable.ic_menu_close_clear_cancel);
            }
            setHasOptionsMenu(true);
            setContentFragment();
            return view;
        }
        //return super.onCreateView(inflater, container, savedInstanceState);
        // Returns the saved view from Dialog Builder on large screens
        return mDialogView;
    }

    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        // Inflate and set the layout for the dialog
        LayoutInflater inflater = getActivity().getLayoutInflater();
        // As it is a Dialog, the root ViewGroup can be null without issues
        @SuppressLint("InflateParams") final View view = inflater.inflate(R.layout.fragment_dynamic_dialog, null);
        setToolbarTitle(view);
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        Bundle args = getArguments();
        // Pass null as the parent view because its going in the dialog layout
        builder.setView(view)
                // Add action buttons
                .setPositiveButton(args.getString("button_positive"), new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        onPositiveButtonClick(view);
                    }
                })
                .setNegativeButton(args.getString("button_negative"), new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        // Send the negative button event back to the host activity
                        // Canceled creation, nothing to do
                        //dialog.cancel();
                        onNegativeButtonClick();
                    }
                });
        if (mHasNeutralButton) {
            builder.setNeutralButton(args.getString("button_neutral"), new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    onNeutralButtonClick(view);
                }
            });
        }
        setContentFragment();
        // Save the View so that it can returned by onCreateView
        // (otherwise it is null and it poses problems when committing child fragment transactions)
        mDialogView = view;
        return builder.create();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        //super.onCreateOptionsMenu(menu, inflater);
        menu.clear();
        getActivity().getMenuInflater().inflate(R.menu.menu_dynamic_fragment, menu);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        // Show the neutral button if needed
        if (!mHasNeutralButton) {
            menu.removeItem(R.id.menu_neutral_button);
        }
        super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Determine which menu item was clicked
        int id = item.getItemId();
        View view = getView();

        // Hide the keyboard if present
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Activity.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(getView().getWindowToken(), 0);
        }
        if (id == R.id.menu_positive_button) {
            // Handle positive button click here
            onPositiveButtonClick(view);
            return true;
        }
        else if (id == R.id.menu_neutral_button) {
            // Handle neutral button click here
            onNeutralButtonClick(view);
            return true;
        }
        else if (id == android.R.id.home) {
            // Handle negative button click here
            onNegativeButtonClick();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDestroyView() {
        Dialog dialog = getDialog();
        // Stop the dialog from being dismissed on rotation, due to a bug with the compatibility library
        // https://code.google.com/p/android/issues/detail?id=17423
        if (dialog != null && getRetainInstance()) {
            dialog.setDismissMessage(null);
        }
        super.onDestroyView();
    }

    private void setContentFragment() {
        // Get the child fragment manager (and not the "normal" one)
        FragmentTransaction transaction = getChildFragmentManager().beginTransaction();

        // Set the actual content of the fragment
        transaction.replace(R.id.dynamic_fragment_content, mContentFragment);

        // Commit the transaction instantly
        transaction.commitNow();
    }

    /** Sets the title of the Fragment from the Tag */
    private Toolbar setToolbarTitle(View view) {
        Toolbar toolbar = (Toolbar) view.findViewById(R.id.dialog_toolbar);
        toolbar.setTitle(getTag());
        return toolbar;
    }

    protected abstract void onPositiveButtonClick(View view);

    protected abstract void onNeutralButtonClick(View view);

    protected abstract void onNegativeButtonClick();
}

File: app/src/main/java/com/wismna/geoffroy/donext/fragments/DynamicDialogFragment.kt
package com.wismna.geoffroy.donext.fragments

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import com.wismna.geoffroy.donext.R
import java.util.Objects

/**
 * Created by wismna on 2017-03-21.
 * Sub-class this class to create a dynamic fragment that will act as a Dialog in large layouts and
 * a full screen fragment in smaller layouts.
 */
abstract class DynamicDialogFragment : DialogFragment() {
    var mButtonCount = 2
    var mPositiveButtonString = ""
    var mNeutralButtonString = ""
    var mNegativeButtonString = ""
    var mContentLayoutId = 0
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // This part is only needed on small layouts (large layouts use onCreateDialog)
        if (!showsDialog) {
            val view = inflater.inflate(R.layout.fragment_dynamic_dialog, container, false)
            val activity = requireActivity() as AppCompatActivity
            //activity.setSupportActionBar(setToolbarTitle(view));
            val actionBar = activity.supportActionBar
            if (actionBar != null) {
                actionBar.setDisplayHomeAsUpEnabled(true)
                actionBar.setHomeButtonEnabled(true)
                actionBar.setHomeAsUpIndicator(android.R.drawable.ic_menu_close_clear_cancel)
            }
            setHasOptionsMenu(true)
            insertContentView(view, inflater)
            return view
        }
        // This basically returns null
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        // Inflate and set the layout for the dialog
        val activity: Activity = requireActivity()
        val inflater = activity.layoutInflater
        // As it is a Dialog, the root ViewGroup can be null without issues
        val view = inflater.inflate(R.layout.fragment_dynamic_dialog, null)
        val builder = AlertDialog.Builder(activity)
        // Add action buttons
        builder.setView(view)
                .setNegativeButton(mNegativeButtonString) { dialog: DialogInterface?, id: Int ->
                    // Send the negative button event back to the host activity
                    // Canceled creation, nothing to do
                    onNegativeButtonClick()
                }
        if (mButtonCount >= 2) {
            builder.setPositiveButton(mPositiveButtonString) { dialog: DialogInterface?, id: Int -> onPositiveButtonClick(view) }
        }
        if (mButtonCount == 3) {
            builder.setNeutralButton(mNeutralButtonString) { dialog: DialogInterface?, which: Int -> onNeutralButtonClick(view) }
        }
        setToolbarTitle(view)
        insertContentView(view, inflater)
        return builder.create()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        requireActivity().menuInflater.inflate(R.menu.menu_dynamic_fragment, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        /*switch (mButtonCount) {
            case 1:
                menu.removeItem(R.id.menu_positive_button);
                menu.removeItem(R.id.menu_neutral_button);
                break;
            case 2:
                menu.removeItem(R.id.menu_neutral_button);
                menu.findItem(R.id.menu_positive_button).setTitle(args.getString("button_positive"));
                break;
            case 3:
                menu.findItem(R.id.menu_neutral_button).setTitle(args.getString("button_neutral"));
                menu.findItem(R.id.menu_positive_button).setTitle(args.getString("button_positive"));
                break;
        }*/

        // Hide buttons depending on count
        when (mButtonCount) {
            1 -> {
                menu.removeItem(R.id.menu_positive_button)
                menu.removeItem(R.id.menu_neutral_button)
            }

            2 -> menu.removeItem(R.id.menu_neutral_button)
        }
        when (mButtonCount) {
            3 -> {
                menu.findItem(R.id.menu_neutral_button).setTitle(mNeutralButtonString)
                menu.findItem(R.id.menu_positive_button).setTitle(mPositiveButtonString)
            }

            2 -> menu.findItem(R.id.menu_positive_button).setTitle(mPositiveButtonString)
        }
        super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Determine which menu item was clicked
        val id = item.itemId
        val view = view

        // Hide the keyboard if present
        if (view != null) {
            val imm = requireActivity().getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(requireView().windowToken, 0)
        }
        when (id) {
            R.id.menu_positive_button -> {
                // Handle positive button click here
                onPositiveButtonClick(view)
                return true
            }
            R.id.menu_neutral_button -> {
                // Handle neutral button click here
                onNeutralButtonClick(view)
                return true
            }
            android.R.id.home -> {
                // Handle negative button click here
                onNegativeButtonClick()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroyView() {
        val dialog = dialog
        // Stop the dialog from being dismissed on rotation, due to a bug with the compatibility library
        // https://code.google.com/p/android/issues/detail?id=17423
        if (dialog != null && retainInstance) {
            dialog.setDismissMessage(null)
        }
        super.onDestroyView()
    }

    fun showFragment(fragmentManager: FragmentManager, title: String?, isLargeLayout: Boolean) {
        if (isLargeLayout) show(fragmentManager, title) else {
            // The device is smaller, so show the fragment fullscreen
            val transaction = fragmentManager.beginTransaction()
            // For a little polish, specify a transition animation
            transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
            // To make it fullscreen, use the 'content' root view as the container
            // for the fragment, which is always the root view for the activity
            transaction.add(android.R.id.content, this, title)
                    .addToBackStack(null).commit()
        }
    }

    /** Helper function to get a View, without having to worry about the fact that is a Dialog or not */
    protected fun <T : View?> findViewById(id: Int): T {
        return if (showsDialog) requireDialog().findViewById(id) else requireView().findViewById(id)
    }

    /** Helper method to clear focus by giving it to the parent layout  */
    protected fun clearFocus() {
        val view = view
        if (view != null) {
            view.requestFocus()

            // Hide keyboard
            val inputMethodManager = requireContext().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            Objects.requireNonNull(inputMethodManager).hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    /** Sets the title of the Fragment from the Tag  */
    private fun setToolbarTitle(view: View): Toolbar {
        val toolbar = view.findViewById<Toolbar>(R.id.dialog_toolbar)
        toolbar.title = tag
        return toolbar
    }

    /** Inserts the actual contents in the content Frame Layout  */
    private fun insertContentView(view: View, inflater: LayoutInflater) {
        // Ensure that the content view is set
        if (mContentLayoutId == 0) return
        // Insert the content view
        val content = view.findViewById<FrameLayout>(R.id.dynamic_fragment_content)
        content.addView(inflater.inflate(mContentLayoutId, view.parent as ViewGroup))
    }

    protected abstract fun onPositiveButtonClick(view: View?)
    protected abstract fun onNeutralButtonClick(view: View?)
    protected abstract fun onNegativeButtonClick()
}
